components/lib/state/src/main/java/mozilla/components/lib/state/ext/ComposeExtensions.kt

```diff
@@ -0,0 +1,117 @@
+
+import android.os.Parcelable
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.ui.platform.LocalLifecycleOwner
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import mozilla.components.lib.state.Action
+import mozilla.components.lib.state.State
+import mozilla.components.lib.state.Store
+import androidx.compose.runtime.State as ComposeState
+
+/**
+ * Starts observing this [Store] and represents the mapped state (using [map]) via [ComposeState].
+ *
+ * Every time the mapped [Store] state changes, the returned [ComposeState] will be updated causing
+ * recomposition of every [ComposeState.value] usage.
+ *
+ * The [Store] observer will automatically be removed when this composable disposes or the current
+ * [LifecycleOwner] moves to the [Lifecycle.State.DESTROYED] state.
+ */
+@Composable
+fun <S : State, A : Action, R> Store<S, A>.observeAsState(map: (S) -> R): ComposeState<R?> {
+    val lifecycleOwner = LocalLifecycleOwner.current
+    val state = remember { mutableStateOf<R?>(map(state)) }
+
+    DisposableEffect(this, lifecycleOwner) {
+        val subscription = observe(lifecycleOwner) { browserState ->
+            state.value = map(browserState)
+        }
+        onDispose { subscription?.unsubscribe() }
+    }
+
+    return state
+}
+
+/**
+ * Starts observing this [Store] and represents the mapped state (using [map]) via [ComposeState].
+ *
+ * Everytime the [Store] state changes and the result of the [observe] function changes for this
+ * state, the returned [ComposeState] will be updated causing recomposition of every
+ * [ComposeState.value] usage.
+ *
+ * The [Store] observer will automatically be removed when this composable disposes or the current
+ * [LifecycleOwner] moves to the [Lifecycle.State.DESTROYED] state.
+ */
+@Composable
+fun <S : State, A : Action, O, R> Store<S, A>.observeAsState(
+    observe: (S) -> O,
+    map: (S) -> R
+): ComposeState<R?> {
+    val lifecycleOwner = LocalLifecycleOwner.current
+    var lastValue = observe(state)
+    val state = remember { mutableStateOf<R?>(map(state)) }
+
+    DisposableEffect(this, lifecycleOwner) {
+        val subscription = observe(lifecycleOwner) { browserState ->
+            val newValue = observe(browserState)
+            if (newValue != lastValue) {
+                state.value = map(browserState)
+                lastValue = newValue
+            }
+        }
+        onDispose { subscription?.unsubscribe() }
+    }
+
+    return state
+}
+
+/**
+ * Helper for creating a [Store] scoped to a `@Composable` and whose [State] gets saved and restored
+ * on process recreation.
+ */
+@Composable
+inline fun <reified S : State, A : Action> scopedSaveableStore(
```

Comment:
Maybe inline with my other suggestion `composableStore`? or `persistentComposableStore`?

Reply:
`Saveable` here comes from the `rememberSaveable` / `Saver` naming in Jetpack Compose and "scoped" since it is scoped to this composable's calling context.
- https://developer.android.com/jetpack/compose/state#restore-ui-state

Reply:
Renaming it to `composableStore()`.